contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,138 @@ The central manager will poll each worker for their discovered volumes and creat
+that is:
+
+1. When the feature gate `DynamicProvisioningScheduling` is enabled,
+the scheduler will verify a pod's volume requirements,
+and set an annotation `annScheduledNode` on PVCs that need provisioning.
+2. New fields `Capacity` will be added in `Status` of `StorageClass` to
+allow storage providers to report capacity limits per topology domain.
+
+The local volume provisioner needs to be updated in several areas to support dynamic provisioning.
+
+##### Backend storage management
+
+Mechanisms to manage local storage sources are varied. They could be LVM, Ceph, or something like that.
+In addition to current `populator`, `deletor` and `discoverer`, a new component named
+`storageManager` will be introduced in the plugin, to manage the backend storage.
+
+`storageManager` is a general interface, each kind of storage should have its own implementation.
+During start-up, a new environment variable `STORAGE_BACKEND` is set in the provisioner DaemonSet,
+to determine which StorageManager instance to start.
+(TBD: Should we support the scenario of multi backends on a node?)
+
+A StorageManager instance is responsible for:
+
+1. Monit changes in storage input, and handle corresponding initialization.
+2. Update capacity in StorageClass when storage source changed.
+3. Provide the ability to create volumes, detailed in the volume creation section.
+
+Take LVM as an example, when a new source disk "/dev/new" is added to its StorageClass `exStorage`,
+the StorageManager should:
+
+1. Check if related PV (physical volume in LVM) exists, if not, create one with command:
+```
+pvcreate /dev/new
+```
+
+2. Check if volume group exStorageGroup exists, if not, create it:
+```
+vgcreate exStorageGroup /dev/new
+```
+
+
+3. If VG exists, check if the physical volume exists in volume group exStorageGroup,
+if not, add it:
+```
+vgextend exStorageGroup /dev/new
+```
+
+4. Update capacity in StorageClass if there is a change in VG size.
+
+A new field `storageSource` will be added into the configuration ConfigMap, to illustrate available
+storage source for provisioning.
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-fast:
+      hostDir: "/mnt/ssds"
+    local-slow:
+      hostDir: "/mnt/hdds"
+    local_dynamic:
+      storageSources:
+        - "/dev/sdb"
+        - "/dev/sdc"
+```
+For example for an LVM StorageManager, the list of storageSource could contain disk partitions.
+For other StorageManagers, the list could contain other storage sources (e.g. unpartitioned disks).
+Each StorageManager should know the scope and what to do with the input.
+
+Major downsides of this approach:
+
+* As provisioners running on all nodes shall a same configuration ConfigMap,
+storage sources of their nodes are required to be configured in the same way.

```

Comment:
When you have a couple of servers with 80 drives each, behind RAID controllers or whatnot, this can be *very* tricky! Any 'mistake' (e.g. due to disk relabeling after a reboot) could cause some... 'interesting' results.